cli/command/service/ps.go

```diff
@@ -43,13 +47,31 @@ func runPS(dockerCli *command.DockerCli, opts psOptions) error {
 	client := dockerCli.Client()
 	ctx := context.Background()
 
-	service, _, err := client.ServiceInspectWithRaw(ctx, opts.serviceID)
-	if err != nil {
-		return err
+	filter := opts.filter.Value()
+	for _, service := range opts.services {
+		serviceIDCount := 0
+		serviceIDFilter := filters.NewArgs()
+		serviceIDFilter.Add("id", service)
+		if serviceList, err := client.ServiceList(ctx, types.ServiceListOptions{Filter: serviceIDFilter}); err == nil {
+			for _, serviceEntry := range serviceList {
+				filter.Add("service", serviceEntry.ID)
+				serviceIDCount++
+			}
+		}
+		serviceNameFilter := filters.NewArgs()
+		serviceNameFilter.Add("name", service)
+		if serviceList, err := client.ServiceList(ctx, types.ServiceListOptions{Filter: serviceNameFilter}); err == nil {
```

Comment:
@dnephin I tried to reduce the number of passes to list the services but cannot find a good way. The issue is that `ServiceList` will not indicate which search term is hit and which search term is a miss. We will not be able to know if any search term returns nothing if we combines search terms into less ServiceList calls.

I also thought about list all services in one call (without filtering), then do the filtering on the client side to find out which search term is a hit and which search term is a miss. This might be OK if the total number of services is small but I am not sure if it will be an issue if the total number of services is big.